Ring allreduce in multitensor

uuuvn opened the pull request:
- [X] Basic ring_allreduce impl
- [ ] Fix last failing test
- [ ] Benchmark to make sure that it's actually faster
- [ ] Hopefully merge

uuuvn commented:
The last failing test is `def test_simple_reduce_1(self): return self._test_simple_reduce_axis(1)`, i suspect that it is because something is wrong with the way i chunk lazybuffers

g1y5x3 commented:
with the `shard` API in tensor, you shouldn't need to do the chunk operations. also, it should support all [reduced ops](https://github.com/tinygrad/tinygrad/blob/9699c8c90b8e4b4d1fa7afd3f86b2cb55d6af3c9/tinygrad/features/multi.py#L69) not be hardcoded to binary add.

uuuvn commented:
Very strange: when i called function with the same LazyBuffers but created from numpy to reproduce & debug everything is ok
Problem also disappears when i add `lbs = [lb.contiguous() for lb in lbs]` at the start of a function to force realize lazybuffers.
Some bugs in lazy/jit/whatever? Maybe i'm doing something that somehow breaks shapetracker?

g1y5x3 commented:
so I was referring to `chunk` in tensor.py but now the question is where is the appropriate place for `chunk`? also, you can refer to `extra/dist/collectives.py` for making it multiprocess. Once I have the multigpu training benchmarks, it would be easier to test the implementation.

uuuvn commented:
> Very strange: when i called function with the same LazyBuffers but created from numpy to reproduce & debug everything is ok Problem also disappears when i add `lbs = [lb.contiguous() for lb in lbs]` at the start of a function to force realize lazybuffers. Some bugs in lazy/jit/whatever? Maybe i'm doing something that somehow breaks shapetracker?

The problem also disappears if i use CLANG or LLVM backend instead of METAL....

uuuvn commented:
So... The only thing left is to make sure that it is actually faster, which at least on my m1 machine isn't the case.

Can somebody run `python test/test_multitensor_speed.py` on a 4+ GPU machine?

geohot commented:
How is the benchmarking coming?

uuuvn commented:
> How is the benchmarking coming?

> Can somebody run python test/test_multitensor_speed.py on a 4+ GPU machine?

Or is benchmark something else than just a python script that outputs results?

geohot commented:
Ahh, running on tiny9

geohot commented:
Hmm, ran a second time, slower
```
tiny@tiny9:~/tinygrad$ python test/test_multitensor_speed.py
Iter (r) 0/10: 0.6794774532318115
Iter (n) 0/10: 0.5218009948730469
Iter (r) 1/10: 0.37977075576782227
Iter (n) 1/10: 0.536362886428833
Iter (r) 2/10: 0.3786904811859131
Iter (n) 2/10: 1.2099571228027344
Iter (r) 3/10: 0.3500063419342041
Iter (n) 3/10: 0.49944448471069336
Iter (r) 4/10: 0.3451883792877197
Iter (n) 4/10: 0.49990272521972656
Iter (r) 5/10: 0.3456103801727295
Iter (n) 5/10: 0.5003464221954346
Iter (r) 6/10: 0.34472084045410156
Iter (n) 6/10: 0.49947381019592285
Iter (r) 7/10: 0.3455815315246582
Iter (n) 7/10: 0.4994392395019531
Iter (r) 8/10: 0.3451573848724365
Iter (n) 8/10: 0.49997830390930176
Iter (r) 9/10: 0.3456544876098633
Iter (n) 9/10: 0.49947142601013184
Average: 0.5766177415847777 naive vs 0.38598580360412593 ring
```

uuuvn commented:
Average: 0.6203446388244629 naive vs 1.1979577302932738 ring

It's actually slower....
0.6 seconds for neive and 1.19 for ring

uuuvn commented:
The first one is old impl - the second one is ring one...

uuuvn commented:
So the first attempt was slower and second was faster. Can you run it a couple more times?

uuuvn commented:
```
tiny@tiny9:~/tinygrad$ python test/test_multitensor_speed.py
Iter (r) 0/10: 8.868210792541504
Iter (n) 0/10: 1.705146074295044
Iter (r) 1/10: 0.35244202613830566
Iter (n) 1/10: 0.49947524070739746
Iter (r) 2/10: 0.3451864719390869
Iter (n) 2/10: 0.4998817443847656
```

It looks like the first one time it was slow, but after that ring was faster

uuuvn commented:
```
Iter (r) 0/10: 8.868210792541504
Iter (n) 0/10: 1.705146074295044
```

That's why on the first run it was slower on average despite after that being faster:
```
Iter (r) 1/10: 0.37977075576782227
Iter (n) 1/10: 0.536362886428833
Iter (r) 2/10: 0.3786904811859131
Iter (n) 2/10: 1.2099571228027344
Iter (r) 3/10: 0.3500063419342041
Iter (n) 3/10: 0.49944448471069336
Iter (r) 4/10: 0.3451883792877197
Iter (n) 4/10: 0.49990272521972656
Iter (r) 5/10: 0.3456103801727295
Iter (n) 5/10: 0.5003464221954346
Iter (r) 6/10: 0.34472084045410156
Iter (n) 6/10: 0.49947381019592285
Iter (r) 7/10: 0.3455815315246582
Iter (n) 7/10: 0.4994392395019531
Iter (r) 8/10: 0.3451573848724365
Iter (n) 8/10: 0.49997830390930176
Iter (r) 9/10: 0.3456544876098633
Iter (n) 9/10: 0.49947142601013184
```

I'll just fix the test to discard the first run

uuuvn commented:
@geohot made fix to discard results from the first iteration which is for some reason much slower.

Can you test it again?

geohot commented:
Your script isn't testing GB/s. What GB/s do you expect? You can see the tinybox bandwidths between GPUs in "AMD Benchmark" CI

uuuvn commented:
> Your script isn't printing GB/s. What GB/s do you expect?
> 
> You can see the tinybox bandwidths between GPUs in "AMD Benchmark" CI

I've rewritten benchmark to output Gb/s and GFLOP/s

uuuvn commented:
Also renamed benchmark script to `test/bench_multitensor_speed.py`, otherwise it's being ran by CI

geohot commented:
```
tiny@tiny9:~/tinygrad$ python3 test/bench_multitensor_speed.py
naive_allreduce iter 0/10: 0.71 sec 9.07 GFLOP/s 108.85 Gb/s
naive_allreduce iter 1/10: 0.65 sec 9.88 GFLOP/s 118.53 Gb/s
naive_allreduce iter 2/10: 0.65 sec 9.89 GFLOP/s 118.71 Gb/s
naive_allreduce iter 3/10: 0.65 sec 9.91 GFLOP/s 118.88 Gb/s
naive_allreduce iter 4/10: 0.65 sec 9.91 GFLOP/s 118.89 Gb/s
naive_allreduce iter 5/10: 0.65 sec 9.89 GFLOP/s 118.69 Gb/s
naive_allreduce iter 6/10: 0.65 sec 9.91 GFLOP/s 118.89 Gb/s
naive_allreduce iter 7/10: 0.65 sec 9.89 GFLOP/s 118.71 Gb/s
naive_allreduce iter 8/10: 0.65 sec 9.89 GFLOP/s 118.71 Gb/s
naive_allreduce iter 9/10: 0.65 sec 9.89 GFLOP/s 118.68 Gb/s
ring_allreduce iter 0/10: 0.45 sec 116.63 GFLOP/s 242.78 Gb/s
ring_allreduce iter 1/10: 0.45 sec 117.74 GFLOP/s 245.10 Gb/s
ring_allreduce iter 2/10: 0.44 sec 118.51 GFLOP/s 246.69 Gb/s
ring_allreduce iter 3/10: 0.44 sec 118.50 GFLOP/s 246.67 Gb/s
ring_allreduce iter 4/10: 0.44 sec 118.38 GFLOP/s 246.42 Gb/s
ring_allreduce iter 5/10: 0.44 sec 118.44 GFLOP/s 246.55 Gb/s
ring_allreduce iter 6/10: 0.44 sec 118.39 GFLOP/s 246.45 Gb/s
ring_allreduce iter 7/10: 0.44 sec 118.46 GFLOP/s 246.60 Gb/s
ring_allreduce iter 8/10: 0.44 sec 118.47 GFLOP/s 246.61 Gb/s
ring_allreduce iter 9/10: 0.44 sec 118.33 GFLOP/s 246.32 Gb/s
Naive:
  0.66 seconds/iter
  9.81 GFLOP/s
  117.75 Gb/s
Ring:
  0.45 seconds/iter
  118.18 GFLOP/s
  246.02 Gb/s
  ```
  
  Gb/s or GB/s. What should it be and what are you getting?

uuuvn commented:
@geohot the only problem was that benchmark recorded time of the first iteration, which was much (4x+) slower, probably due to compiling kernels. I've rewritten benchmark to account for this and print GB/s and renamed from `test/test_multitensor_speed.py` to `test/bench_multitensor_speed.py`, otherwise it's being run by CI expecting it to be a unit test.

uuuvn commented:
Changed it to GB/s, didn't thought that in english GB/s and Gb/s are gigabytes/gigabits.

uuuvn commented:
It looks like in theory it should copy 2K(N - 1) where K is size of each lazybuffer and N is number of nodes/gpus

uuuvn commented:
That's only copies between gpus/nodes, not counting stuff that happens inside one gpu

uuuvn commented:
There should be only about 6.44 GB copied between GPUs. The problem is that there is all sorts of copies apart from those (eg copies from one device to others while sharding tensor) counted

chenyuxyz commented on test/test_multitensor.py:
```diff
@@ -69,6 +74,38 @@ def _test_simple_reduce_axis(self, shard_x):
     O = X.sum(axis=1)
     np.testing.assert_allclose(O.numpy(), 256)
 
+  def test_fuzz_allreduce(self):
```
this is quite slow on CLANG backend, took 9 minutes to finish. not sure if it's expected

uuuvn replied:
That's not expected - will investigate why

uuuvn replied:
```
Program: /Users/user/src/tinygrad/.venv/bin/pyinstrument wtf.py

27.106 <module>  wtf.py:1
└─ 26.899 Tensor.numpy  tinygrad/tensor.py:143
   └─ 26.892 Tensor.realize  tinygrad/tensor.py:112
      └─ 26.767 run_schedule  tinygrad/realize.py:25
         └─ 26.587 lower_schedule_item  tinygrad/realize.py:16
            └─ 26.526 Compiled.get_runner  tinygrad/device.py:329
               ├─ 25.291 Compiled.to_program  tinygrad/device.py:292
               │  ├─ 23.976 CompiledASTRunner.build  tinygrad/device.py:261
               │  │  └─ 23.975 ClangProgram.__init__  tinygrad/runtime/ops_clang.py:16
               │  │     └─ 23.633 CDLL.__init__  ctypes/__init__.py:342
               │  │           [2 frames hidden]  ctypes, <built-in>
               │  │              23.620 dlopen  <built-in>
               │  └─ 1.096 Linearizer.linearize  tinygrad/codegen/linearizer.py:181
               │     └─ 0.432 <dictcomp>  tinygrad/codegen/linearizer.py:383
               │        └─ 0.412 Linearizer.global_load  tinygrad/codegen/linearizer.py:85
               └─ 1.224 Compiled.get_linearizer  tinygrad/device.py:305
                  ├─ 0.890 Linearizer.__init__  tinygrad/codegen/kernel.py:70
                  │  ├─ 0.436 get_lazyop_info  tinygrad/ops.py:95
                  │  │  └─ 0.425 run_ast  tinygrad/ops.py:97
                  │  │     └─ 0.410 <listcomp>  tinygrad/ops.py:98
                  │  │        └─ 0.410 run_ast  tinygrad/ops.py:97
                  │  │           └─ 0.401 <listcomp>  tinygrad/ops.py:98
                  │  │              └─ 0.401 run_ast  tinygrad/ops.py:97
                  │  │                 └─ 0.360 <listcomp>  tinygrad/ops.py:98
                  │  │                    └─ 0.360 run_ast  tinygrad/ops.py:97
                  │  │                       └─ 0.330 <listcomp>  tinygrad/ops.py:98
                  │  │                          └─ 0.329 run_ast  tinygrad/ops.py:97
                  │  │                             └─ 0.281 <listcomp>  tinygrad/ops.py:98
                  │  │                                └─ 0.281 run_ast  tinygrad/ops.py:97
                  │  └─ 0.283 Linearizer.simplify_merge_adjacent  tinygrad/codegen/kernel.py:249
                  └─ 0.325 Linearizer.hand_coded_optimizations  tinygrad/codegen/kernel.py:471
```

It looks like it wastes a ton of time in dlopen()

uuuvn replied:
![image](https://github.com/tinygrad/tinygrad/assets/83587632/552e3a86-2ea2-46ac-ac48-37d8311ed6cf)
Reproduced with just adding two tensors. It looks like for some reason dlopen in python is horribly slow, thus high kernel dispatch overhead, thus it takes ages when there is a lot of small computations (like in fuzzing stuff)

I can just disable it for clang, but that doesn't feel like a right fix

uuuvn replied:
Searched for how other fuzzing is done in tinygrad - it is in separate `tests/external` directory that isn't being ran by CI, should i just move my fuzzing here?

uuuvn replied:
Moved both fuzzing and benchmark to external

chenyuxyz replied:
symbolic and shapetracker fuzzer runs under a different worker and those are device agnostic

for testing multitensor allreduce, i think we want at least one interpreted backend and one compiled backend. fine to skip for clang if it's the only slow one due to writing to files

uuuvn replied:
Done

chenyuxyz commented:
on tiny9, the number does not look right

```
(venv) ➜  tinygrad git:(3000) ✗ python test/external/external_benchmark_multitensor_allreduce.py
naive_allreduce iter 0/10: 0.00 sec 1909.11 GFLOP/s 22909.37 GB/s
naive_allreduce iter 1/10: 0.00 sec 1979.02 GFLOP/s 23748.29 GB/s
naive_allreduce iter 2/10: 0.00 sec 1958.37 GFLOP/s 23500.45 GB/s
naive_allreduce iter 3/10: 0.00 sec 2037.98 GFLOP/s 24455.78 GB/s
naive_allreduce iter 4/10: 0.00 sec 1982.80 GFLOP/s 23793.60 GB/s
naive_allreduce iter 5/10: 0.00 sec 2023.33 GFLOP/s 24279.98 GB/s
naive_allreduce iter 6/10: 0.00 sec 1979.75 GFLOP/s 23756.99 GB/s
naive_allreduce iter 7/10: 0.00 sec 2046.01 GFLOP/s 24552.07 GB/s
naive_allreduce iter 8/10: 0.00 sec 1986.44 GFLOP/s 23837.33 GB/s
naive_allreduce iter 9/10: 0.00 sec 2045.70 GFLOP/s 24548.35 GB/s
ring_allreduce iter 0/10: 0.01 sec 1037.09 GFLOP/s 11830.55 GB/s
ring_allreduce iter 1/10: 0.01 sec 1041.43 GFLOP/s 11880.00 GB/s
ring_allreduce iter 2/10: 0.01 sec 1051.12 GFLOP/s 11990.50 GB/s
ring_allreduce iter 3/10: 0.01 sec 1058.21 GFLOP/s 12071.47 GB/s
ring_allreduce iter 4/10: 0.01 sec 1058.14 GFLOP/s 12070.63 GB/s
ring_allreduce iter 5/10: 0.01 sec 990.40 GFLOP/s 11297.88 GB/s
ring_allreduce iter 6/10: 0.01 sec 1065.86 GFLOP/s 12158.66 GB/s
ring_allreduce iter 7/10: 0.01 sec 1058.66 GFLOP/s 12076.52 GB/s
ring_allreduce iter 8/10: 0.01 sec 1067.20 GFLOP/s 12174.03 GB/s
ring_allreduce iter 9/10: 0.01 sec 1057.73 GFLOP/s 12066.01 GB/s
Naive:
  0.00 seconds/iter
  1994.85 GFLOP/s
  23938.22 GB/s
Ring:
  0.01 seconds/iter
  1048.58 GFLOP/s
  11961.63 GB/s
```

uuuvn commented:
Numbers are ok on metal, i've updated benchmark to always use time.time() instead of `GlobalCounters.time_sum_s` if it's > 0 and time.time() otherwise. Is it running that fast or is it just outputting near-zero as time but running okay?

chenyuxyz commented:
synced to `6a3b701` and included wall time

```
➜  tinygrad git:(3000) ✗ time python test/external/external_benchmark_multitensor_allreduce.py
naive_allreduce iter 0/10: 0.00 sec 1894.92 GFLOP/s 22739.07 GB/s
naive_allreduce iter 1/10: 0.00 sec 1979.60 GFLOP/s 23755.25 GB/s
naive_allreduce iter 2/10: 0.00 sec 1949.89 GFLOP/s 23398.70 GB/s
naive_allreduce iter 3/10: 0.00 sec 2008.29 GFLOP/s 24099.53 GB/s
naive_allreduce iter 4/10: 0.00 sec 1975.26 GFLOP/s 23703.16 GB/s
naive_allreduce iter 5/10: 0.00 sec 2005.46 GFLOP/s 24065.55 GB/s
naive_allreduce iter 6/10: 0.00 sec 1971.66 GFLOP/s 23659.92 GB/s
naive_allreduce iter 7/10: 0.00 sec 2015.78 GFLOP/s 24189.42 GB/s
naive_allreduce iter 8/10: 0.00 sec 1978.59 GFLOP/s 23743.07 GB/s
naive_allreduce iter 9/10: 0.00 sec 2039.52 GFLOP/s 24474.24 GB/s
ring_allreduce iter 0/10: 0.01 sec 1003.41 GFLOP/s 11446.30 GB/s
ring_allreduce iter 1/10: 0.01 sec 1013.17 GFLOP/s 11557.70 GB/s
ring_allreduce iter 2/10: 0.01 sec 1033.36 GFLOP/s 11787.92 GB/s
ring_allreduce iter 3/10: 0.01 sec 1041.11 GFLOP/s 11876.34 GB/s
ring_allreduce iter 4/10: 0.01 sec 1036.88 GFLOP/s 11828.13 GB/s
ring_allreduce iter 5/10: 0.01 sec 993.47 GFLOP/s 11332.96 GB/s
ring_allreduce iter 6/10: 0.01 sec 1046.20 GFLOP/s 11934.38 GB/s
ring_allreduce iter 7/10: 0.01 sec 1035.19 GFLOP/s 11808.80 GB/s
ring_allreduce iter 8/10: 0.01 sec 1033.67 GFLOP/s 11791.53 GB/s
ring_allreduce iter 9/10: 0.01 sec 1023.41 GFLOP/s 11674.43 GB/s
Naive:
  0.00 seconds/iter
  1981.90 GFLOP/s
  23782.79 GB/s
Ring:
  0.01 seconds/iter
  1025.99 GFLOP/s
  11703.85 GB/s
python test/external/external_benchmark_multitensor_allreduce.py  3.23s user 5.79s system 486% cpu 1.853 total
```

uuuvn commented:
Rebased on master

It is ready for review

chenyuxyz commented:
running `HALF=1 GPUS=2 BS=1024 STEPS=500 python3 examples/hlb_cifar10.py` failed at the end, with 4 gpus it failed at the jit earlier too
```
495  134.60 ms run,   11.25 ms python,  123.34 ms HIP * 2,  479.75 loss, 0.000120 LR, 6.62 GB used,  10031.14 GFLOPS,   1350.16 GOPS
496  133.77 ms run,   11.21 ms python,  122.55 ms HIP * 2,  480.50 loss, 0.000112 LR, 6.62 GB used,  10093.28 GFLOPS,   1350.16 GOPS
497  135.11 ms run,   11.29 ms python,  123.82 ms HIP * 2,  477.75 loss, 0.000102 LR, 6.63 GB used,   9992.93 GFLOPS,   1350.16 GOPS
498  134.18 ms run,   11.36 ms python,  122.82 ms HIP * 2,  482.00 loss, 0.000094 LR, 6.63 GB used,  10062.04 GFLOPS,   1350.16 GOPS
499  136.82 ms run,   11.23 ms python,  125.59 ms HIP * 2,  481.75 loss, 0.000086 LR, 6.63 GB used,   9868.10 GFLOPS,   1350.16 GOPS
shuffling test dataset in 43.35 ms (epoch=0)
Traceback (most recent call last):
  File "/home/chenyu/tinygrad/examples/hlb_cifar10.py", line 382, in <module>
    train_cifar()
  File "/home/chenyu/tinygrad/examples/hlb_cifar10.py", line 340, in train_cifar
    correct, loss = eval_step_jitted(model, Xt, Yt)
  File "/home/chenyu/tinygrad/tinygrad/jit.py", line 115, in __call__
    for ji in self.jit_cache: ji.prg(cast(List[Buffer], ji.rawbufs), var_vals, wait=DEBUG>=2, jit=True)
  File "/home/chenyu/tinygrad/tinygrad/device.py", line 309, in __call__
    et = self.clprg(*[x._buf for x in rawbufs], **lra, vals=tuple(var_vals[k] for k in self.vars), wait=wait or DEBUG>=2)
  File "/home/chenyu/tinygrad/tinygrad/runtime/ops_hip.py", line 64, in __call__
    check(hip.hipModuleLaunchKernel(self.prg, *global_size, *local_size, 0, None, None, self.vargs))
  File "/home/chenyu/tinygrad/tinygrad/runtime/ops_hip.py", line 31, in check
    if status != 0: raise RuntimeError(f"HIP Error {status}, {ctypes.string_at(hip.hipGetErrorString(status)).decode()}")
RuntimeError: HIP Error 101, invalid device ordinal
HALF=1 GPUS=2 BS=1024 STEPS=500 python3 examples/hlb_cifar10.py  82.71s user 14.52s system 107% cpu 1:30.43 total
```

also the step time is a lot slower than the one on master
```
(venv) ➜  tinygrad git:(master) ✗ time HALF=1 GPUS=2 BS=1024 STEPS=10 python3 examples/hlb_cifar10.py
shuffling training dataset in 836.26 ms (epoch=0)
  0 5641.08 ms run, 5630.02 ms python,   11.06 ms HIP * 2, 1198.00 loss, 0.001493 LR, 0.46 GB used,    239.53 GFLOPS,   1351.21 GOPS
  1  362.36 ms run,  358.39 ms python,    3.97 ms HIP * 2, 1193.00 loss, 0.002987 LR, 5.07 GB used,   3726.03 GFLOPS,   1350.17 GOPS
  2  104.85 ms run,   12.45 ms python,   92.40 ms HIP * 2, 1614.00 loss, 0.003132 LR, 5.08 GB used,  12877.15 GFLOPS,   1350.17 GOPS
  3  104.51 ms run,   11.64 ms python,   92.87 ms HIP * 2, 2478.00 loss, 0.002697 LR, 5.08 GB used,  12919.10 GFLOPS,   1350.17 GOPS
  4  103.27 ms run,   11.56 ms python,   91.71 ms HIP * 2, 2364.00 loss, 0.002262 LR, 5.08 GB used,  13074.74 GFLOPS,   1350.17 GOPS
  5  103.56 ms run,   11.51 ms python,   92.05 ms HIP * 2, 1826.00 loss, 0.001827 LR, 5.09 GB used,  13037.90 GFLOPS,   1350.17 GOPS
  6  103.00 ms run,   11.47 ms python,   91.54 ms HIP * 2, 1446.00 loss, 0.001391 LR, 5.09 GB used,  13108.08 GFLOPS,   1350.17 GOPS
  7  104.00 ms run,   11.54 ms python,   92.46 ms HIP * 2, 1297.00 loss, 0.000957 LR, 5.09 GB used,  12982.21 GFLOPS,   1350.17 GOPS
  8  104.16 ms run,   11.62 ms python,   92.54 ms HIP * 2, 1279.00 loss, 0.000521 LR, 5.10 GB used,  12962.55 GFLOPS,   1350.17 GOPS
  9  104.41 ms run,   11.43 ms python,   92.98 ms HIP * 2, 1288.00 loss, 0.000086 LR, 5.10 GB used,  12931.13 GFLOPS,   1350.17 GOPS
shuffling test dataset in 44.55 ms (epoch=0)
eval     1486/10000 14.86%,    2.45 val_loss STEP=10 (in 1857.53 ms)
HALF=1 GPUS=2 BS=1024 STEPS=10 python3 examples/hlb_cifar10.py  13.40s user 7.02s system 151% cpu 13.442 total
```

uuuvn commented:
> running `HALF=1 GPUS=2 BS=1024 STEPS=500 python3 examples/hlb_cifar10.py` failed at the end, with 4 gpus it failed at the jit earlier too
> 
> ```
> 495  134.60 ms run,   11.25 ms python,  123.34 ms HIP * 2,  479.75 loss, 0.000120 LR, 6.62 GB used,  10031.14 GFLOPS,   1350.16 GOPS
> 496  133.77 ms run,   11.21 ms python,  122.55 ms HIP * 2,  480.50 loss, 0.000112 LR, 6.62 GB used,  10093.28 GFLOPS,   1350.16 GOPS
> 497  135.11 ms run,   11.29 ms python,  123.82 ms HIP * 2,  477.75 loss, 0.000102 LR, 6.63 GB used,   9992.93 GFLOPS,   1350.16 GOPS
> 498  134.18 ms run,   11.36 ms python,  122.82 ms HIP * 2,  482.00 loss, 0.000094 LR, 6.63 GB used,  10062.04 GFLOPS,   1350.16 GOPS
> 499  136.82 ms run,   11.23 ms python,  125.59 ms HIP * 2,  481.75 loss, 0.000086 LR, 6.63 GB used,   9868.10 GFLOPS,   1350.16 GOPS
> shuffling test dataset in 43.35 ms (epoch=0)
> Traceback (most recent call last):
>   File "/home/chenyu/tinygrad/examples/hlb_cifar10.py", line 382, in <module>
>     train_cifar()
>   File "/home/chenyu/tinygrad/examples/hlb_cifar10.py", line 340, in train_cifar
>     correct, loss = eval_step_jitted(model, Xt, Yt)
>   File "/home/chenyu/tinygrad/tinygrad/jit.py", line 115, in __call__
>     for ji in self.jit_cache: ji.prg(cast(List[Buffer], ji.rawbufs), var_vals, wait=DEBUG>=2, jit=True)
>   File "/home/chenyu/tinygrad/tinygrad/device.py", line 309, in __call__
>     et = self.clprg(*[x._buf for x in rawbufs], **lra, vals=tuple(var_vals[k] for k in self.vars), wait=wait or DEBUG>=2)
>   File "/home/chenyu/tinygrad/tinygrad/runtime/ops_hip.py", line 64, in __call__
>     check(hip.hipModuleLaunchKernel(self.prg, *global_size, *local_size, 0, None, None, self.vargs))
>   File "/home/chenyu/tinygrad/tinygrad/runtime/ops_hip.py", line 31, in check
>     if status != 0: raise RuntimeError(f"HIP Error {status}, {ctypes.string_at(hip.hipGetErrorString(status)).decode()}")
> RuntimeError: HIP Error 101, invalid device ordinal
> HALF=1 GPUS=2 BS=1024 STEPS=500 python3 examples/hlb_cifar10.py  82.71s user 14.52s system 107% cpu 1:30.43 total
> ```
> 
> also the step time is a lot slower than the one on master
> 
> ```
> (venv) ➜  tinygrad git:(master) ✗ time HALF=1 GPUS=2 BS=1024 STEPS=10 python3 examples/hlb_cifar10.py
> shuffling training dataset in 836.26 ms (epoch=0)
>   0 5641.08 ms run, 5630.02 ms python,   11.06 ms HIP * 2, 1198.00 loss, 0.001493 LR, 0.46 GB used,    239.53 GFLOPS,   1351.21 GOPS
>   1  362.36 ms run,  358.39 ms python,    3.97 ms HIP * 2, 1193.00 loss, 0.002987 LR, 5.07 GB used,   3726.03 GFLOPS,   1350.17 GOPS
>   2  104.85 ms run,   12.45 ms python,   92.40 ms HIP * 2, 1614.00 loss, 0.003132 LR, 5.08 GB used,  12877.15 GFLOPS,   1350.17 GOPS
>   3  104.51 ms run,   11.64 ms python,   92.87 ms HIP * 2, 2478.00 loss, 0.002697 LR, 5.08 GB used,  12919.10 GFLOPS,   1350.17 GOPS
>   4  103.27 ms run,   11.56 ms python,   91.71 ms HIP * 2, 2364.00 loss, 0.002262 LR, 5.08 GB used,  13074.74 GFLOPS,   1350.17 GOPS
>   5  103.56 ms run,   11.51 ms python,   92.05 ms HIP * 2, 1826.00 loss, 0.001827 LR, 5.09 GB used,  13037.90 GFLOPS,   1350.17 GOPS
>   6  103.00 ms run,   11.47 ms python,   91.54 ms HIP * 2, 1446.00 loss, 0.001391 LR, 5.09 GB used,  13108.08 GFLOPS,   1350.17 GOPS
>   7  104.00 ms run,   11.54 ms python,   92.46 ms HIP * 2, 1297.00 loss, 0.000957 LR, 5.09 GB used,  12982.21 GFLOPS,   1350.17 GOPS
>   8  104.16 ms run,   11.62 ms python,   92.54 ms HIP * 2, 1279.00 loss, 0.000521 LR, 5.10 GB used,  12962.55 GFLOPS,   1350.17 GOPS
>   9  104.41 ms run,   11.43 ms python,   92.98 ms HIP * 2, 1288.00 loss, 0.000086 LR, 5.10 GB used,  12931.13 GFLOPS,   1350.17 GOPS
> shuffling test dataset in 44.55 ms (epoch=0)
> eval     1486/10000 14.86%,    2.45 val_loss STEP=10 (in 1857.53 ms)
> HALF=1 GPUS=2 BS=1024 STEPS=10 python3 examples/hlb_cifar10.py  13.40s user 7.02s system 151% cpu 13.442 total
> ```

Second problem was likely caused by the fact that shapes in allreduce in that benchmark look something like (1, 1, 1, 512, 1), I fixed it by flattening everything to 1d and reshaping back after everything is done, not sure if that's the best way to do it though.

But I've no idea what might have caused the first one at all. `RuntimeError: HIP Error 101, invalid device ordinal`. I only "change" devices with copy_to_device and the destination device is taken from another lazybuffer. The fact that it fails with JIT earlier made me look into what jit does with devices with no success.

uuuvn commented:
Everything should be fixed

chenyuxyz commented on tinygrad/features/multi.py:
```diff
@@ -1,16 +1,38 @@
-  # TODO: replace this with ring reduce
+def ring_allreduce(op: ReduceOps, lbs: List[LazyBuffer]):
+  assert all_int(lbs[0].shape), f"does not support symbolic shape {lbs[0].shape}"
+  assert all_same([lb.shape[0] for lb in lbs]), "allreduce with uneven shards is undefined"
```
why is this not defined?

uuuvn replied:
![image](https://github.com/tinygrad/tinygrad/assets/83587632/995473c8-9835-4461-b78c-e82c57ec9e82)
Because elementwise adds with different shapes are undefined? I thought about just padding with zeroes but because + op in tinygrad doesn't do this i decided that i shouldn't do either.

chenyuxyz commented on tinygrad/features/multi.py:
```diff
@@ -1,16 +1,38 @@
+  n_lbs, dim = len(lbs), prod(lbs[0].shape)
+  base, left = dim // n_lbs, dim % n_lbs
+  c_lens = [base + 1 if left - i > 0 else base for i in range(n_lbs)]
+  acc = 0; chunks = [(acc, (acc := acc + i)) for i in c_lens if i > 0] # noqa: E702
```
don't concat lines like this

uuuvn replied:
fixed

uuuvn replied:
Btw i'm not sure should i resolve conversations like this after fixing them or let you close them after reviewing the fix?

chenyuxyz commented on test/external/external_benchmark_multitensor_allreduce.py:
```diff
@@ -0,0 +1,56 @@
+  return [functools.reduce(lambda x,y: x.e(bop, y), [x.copy_to_device(lb.device) for x in lbs]) for lb in lbs]
+
+def realize(x: LazyBuffer | List[LazyBuffer]):
+    run_schedule(create_schedule(x if isinstance(x, list) else [x]))
```
indent

uuuvn replied:
fixed

chenyuxyz commented on test/external/external_benchmark_multitensor_allreduce.py:
```diff
@@ -0,0 +1,56 @@
+  secs, gflops, gbs = 0, 0, 0
+  for i in range(-2, iters):
+    GlobalCounters.reset()
+    lbs = [LazyBuffer.loadop(LoadOps.CONST, (), dtypes.float, d, float(1+i)).reshape((1,)).expand((N,)).e(LoadOps.CONTIGUOUS) for i,d in enumerate(devs)] # noqa: E501
```
can you use the shard API to construct the sharded tensor? I think it's more useful to test the construction using user facing api. Same for the unit test in test_multitensor

uuuvn replied:
It first creates it on some device and copies to all other devices afterwards which is slower + will eat more memory, with this it creates stuff directly on device where it should be.

uuuvn replied:
I can replace `LazyBuffer.loadop(LoadOps.CONST, (), dtypes.float, d, float(1+i)).reshape((1,)).expand((N,)).e(LoadOps.CONTIGUOUS)` with something like `Tensor.full((N,), float(1+i), device=d).contiguous()` though

uuuvn replied:
> Same for the unit test in test_multitensor

```python
t = Tensor.rand(tuple([(n if i == 0 else 1) * randint(1, 10) for i in range(randint(1, 4))])).shard_(tuple([d0, d1, d2, d3][:n]), 0)
```
I do it with shard here (because size is small and it doesn't matter here)

chenyuxyz commented:
it's faster for hlb_cifar. On tinybox with HSA backend, `STEPS=10 GPUS=6 BS=510 HSA=1 python3 examples/hlb_cifar10.py`
master
```
shuffling training dataset in 1306.38 ms (epoch=0)
  0 14542.83 ms run, 14538.90 ms python,    3.93 ms HSA * 6, 1198.51 loss, 0.001495 LR, 1.10 GB used,     46.18 GFLOPS,    671.61 GOPS
  1 1236.23 ms run, 1234.45 ms python,    1.79 ms HSA * 6, 1197.36 loss, 0.002989 LR, 5.40 GB used,    541.77 GFLOPS,    669.76 GOPS
  2   44.17 ms run,    9.22 ms python,   34.95 ms HSA * 6, 1930.62 loss, 0.003133 LR, 5.41 GB used,  15162.84 GFLOPS,    669.76 GOPS
  3   35.05 ms run,    7.55 ms python,   27.51 ms HSA * 6, 2812.77 loss, 0.002698 LR, 5.41 GB used,  19107.75 GFLOPS,    669.76 GOPS
  4   34.63 ms run,    7.47 ms python,   27.15 ms HSA * 6, 2074.46 loss, 0.002262 LR, 5.41 GB used,  19342.62 GFLOPS,    669.76 GOPS
  5   35.25 ms run,    7.51 ms python,   27.74 ms HSA * 6, 2112.30 loss, 0.001827 LR, 5.41 GB used,  19000.57 GFLOPS,    669.76 GOPS
  6   34.93 ms run,    7.44 ms python,   27.50 ms HSA * 6, 2035.17 loss, 0.001392 LR, 5.41 GB used,  19171.99 GFLOPS,    669.76 GOPS
  7   35.02 ms run,    7.42 ms python,   27.60 ms HSA * 6, 1544.58 loss, 0.000956 LR, 5.41 GB used,  19127.67 GFLOPS,    669.76 GOPS
  8   34.96 ms run,    7.45 ms python,   27.51 ms HSA * 6, 1582.97 loss, 0.000521 LR, 5.41 GB used,  19156.23 GFLOPS,    669.76 GOPS
  9   34.93 ms run,    7.46 ms python,   27.47 ms HSA * 6, 1529.48 loss, 0.000086 LR, 5.41 GB used,  19175.55 GFLOPS,    669.76 GOPS
```
this
```
shuffling training dataset in 1303.38 ms (epoch=0)
  0 15102.22 ms run, 15097.98 ms python,    4.24 ms HSA * 6, 1198.51 loss, 0.001495 LR, 1.10 GB used,     44.47 GFLOPS,    671.63 GOPS
  1 1352.36 ms run, 1350.23 ms python,    2.13 ms HSA * 6, 1197.36 loss, 0.002989 LR, 5.16 GB used,    495.27 GFLOPS,    669.78 GOPS
  2   40.80 ms run,   10.47 ms python,   30.33 ms HSA * 6, 1930.62 loss, 0.003133 LR, 5.17 GB used,  16415.25 GFLOPS,    669.78 GOPS
  3   32.18 ms run,    8.99 ms python,   23.19 ms HSA * 6, 2812.77 loss, 0.002698 LR, 5.17 GB used,  20813.57 GFLOPS,    669.78 GOPS
  4   31.94 ms run,    8.97 ms python,   22.97 ms HSA * 6, 2074.45 loss, 0.002262 LR, 5.17 GB used,  20966.99 GFLOPS,    669.78 GOPS
  5   31.71 ms run,    8.98 ms python,   22.72 ms HSA * 6, 2112.22 loss, 0.001827 LR, 5.17 GB used,  21123.61 GFLOPS,    669.78 GOPS
  6   31.81 ms run,    9.04 ms python,   22.77 ms HSA * 6, 2035.44 loss, 0.001392 LR, 5.17 GB used,  21056.84 GFLOPS,    669.78 GOPS
  7   31.97 ms run,    8.85 ms python,   23.12 ms HSA * 6, 1544.70 loss, 0.000956 LR, 5.17 GB used,  20951.94 GFLOPS,    669.78 GOPS
  8   33.08 ms run,    9.90 ms python,   23.18 ms HSA * 6, 1583.13 loss, 0.000521 LR, 5.17 GB used,  20248.99 GFLOPS,    669.78 GOPS
  9   31.97 ms run,    9.01 ms python,   22.97 ms HSA * 6, 1525.04 loss, 0.000086 LR, 5.17 GB used,  20947.07 GFLOPS,    669.78 GOPS
```

but llama is slower, `HSA=1 python3 examples/llama.py  --temperature=0 --count=10 --prompt="Hello." --shard 2 --size 7B --timing`
master
```
enqueue in   7.03 ms
total  39.27 ms, 25.46 tok/sec
 year
enqueue in   6.96 ms
total  39.14 ms, 25.55 tok/sec
 old
enqueue in   6.94 ms
total  38.99 ms, 25.65 tok/sec
 male
```

this
```
enqueue in   9.17 ms
total  49.42 ms, 20.23 tok/sec
 year
enqueue in   9.14 ms
total  49.40 ms, 20.24 tok/sec
 old
enqueue in   9.11 ms
total  49.50 ms, 20.20 tok/sec
 male
```

It might be ring allreduce or HSA issue (enqueue time increased too). @nimlgen fyi

Given that it's not always faster, we don't merge it as is. Either fixing that, or if we can understand this better (having a script that reproduced the case when ring allreduce is slower), we can merge this with a flag to switch between allreduce and ring allreduce.

chenyuxyz commented:
Is this good to go? Can you post some llama numbers. Test conflicted with another one that I just merged.

Also consider not always squash, I find it difficult to track the diff of these 100+ lines pr, and having multiple commit helps. We always squash the pr as a whole when merging.

uuuvn commented:
Not ready yet

uuuvn commented:
<img width="551" alt="image" src="https://github.com/tinygrad/tinygrad/assets/83587632/7c72df3d-8324-43fd-b847-0010736cff0a">

There is a bunch of problems with combination of JIT and HSA, debugging them now. After that will make LLaMA faster

uuuvn commented:
Depends on #3560 and #3575

uuuvn commented:
First version of `ScheduleBarrier`'s:
```
(.venv) uuuvn@tiny10:~/src/tinygrad$ HSA=1 python3 examples/llama.py  --temperature=0 --count=10 --prompt="Hello." --shard 2 --size 7B --timing
using HSA backend
using LLaMA-7B model
ram used: 13.48 GB, freqs_cis                                         : 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 292/292 [00:05<00:00, 53.30it/s]
loaded weights in 5482.83 ms, 13.48 GB loaded at 2.46 GB/s
Hello.
enqueue in 2430.78 ms
total 2433.90 ms, 0.41 tok/sec
 I
enqueue in 1944.33 ms
total 1946.84 ms, 0.51 tok/sec
'
enqueue in 1854.69 ms
total 1857.27 ms, 0.54 tok/sec
m
enqueue in   9.63 ms
total  52.52 ms, 19.04 tok/sec
 a
enqueue in   9.26 ms
total  44.57 ms, 22.44 tok/sec

enqueue in   9.34 ms
total  44.11 ms, 22.67 tok/sec
2
enqueue in   9.26 ms
total  44.39 ms, 22.53 tok/sec
0
enqueue in   9.28 ms
total  44.34 ms, 22.55 tok/sec
 year
enqueue in   9.28 ms
total  44.42 ms, 22.51 tok/sec
 old
enqueue in   9.23 ms
total  44.55 ms, 22.45 tok/sec
 male
output validated
(.venv) uuuvn@tiny10:~/src/tinygrad$
```
LLaMA 44.55 down from 49.50 w/o ScheduleBarrier (38.99 master)
hlb_cifar +-same

nimlgen commented:
btw I just pushed my old experiments with scheduling within one graph (maybe that's not the best place to do this, but the fastest to hack). The idea is to reschedule to hide copies time  https://github.com/nimlgen/tinygrad/commit/f45c84be6857f2becfd634c85b07cc09ddc54898. Just rebased and retested them, got `total  42.95 ms, 23.29 tok/sec` for llama 7b gen1 on 2gpus. Looks like you doing the same thing on another level? Maybe not barriers, but some heuristics, could it be better?

uuuvn commented:
Ready for review, but #3560 and #3643 are required before merge

uuuvn commented:
> Is this good to go? Can you post some llama numbers. Test conflicted with another one that I just merged.

`HSA=1 python3 examples/llama.py  --temperature=0 --count=10 --prompt="Hello." --shard X --size 7B --timing`
2 GPUs: 34.61 ms => 33.98 ms | -1.82%
3 GPUs: 114.05 ms => 92.06 ms | -19.28%
4 GPUs: 80.54 ms => 54.76 ms | -32.00%
5 GPUs: 107.93 ms => 66.78 ms | -38.12%
6 GPUs: 190.57 ms => 89.29 ms | -53.14%
`STEPS=10 GPUS=X BS=720 HSA=1 python3 examples/hlb_cifar10.py` 
2 GPUs 94.00 ms => 93.24 ms | -0.80%
3 GPUs 61.70 ms => 60.84 ms | -1.39%
4 GPUs 54.93 ms => 53.39 ms | -2.80%
5 GPUs 46.90 ms => 43.16 ms | -7.97%
6 GPUs 69.34 ms => 61.68 ms | -11.04%

chenyuxyz commented:
@nimlgen can you review this one and see how much it overlaps with #3626 and which approach (or both) we will keep long term

geohot commented:
Can confirm it improves speed of cifar a little
```
tiny@tiny12:~/tinygrad$ HSA=1 HALF=1 STEPS=350 BS=1536 GPUS=6 TARGET_EVAL_ACC_PCT=93 python3 examples/hlb_cifar10.py
shuffling training dataset in 958.98 ms (epoch=0)
  0 18045.31 ms run, 18039.51 ms python,    5.80 ms HSA * 6, 1197.00 loss, 0.000043 LR, 0.56 GB used,    111.74 GFLOPS,   2016.42 GOPS
  1 1139.61 ms run, 1136.97 ms python,    2.65 ms HSA * 6, 1195.00 loss, 0.000085 LR, 6.63 GB used,   1767.76 GFLOPS,   2014.56 GOPS
  2   53.81 ms run,   13.86 ms python,   39.96 ms HSA * 6, 1178.00 loss, 0.000128 LR, 6.64 GB used,  37436.16 GFLOPS,   2014.56 GOPS
  3   45.62 ms run,   12.35 ms python,   33.27 ms HSA * 6, 1163.00 loss, 0.000171 LR, 6.64 GB used,  44158.40 GFLOPS,   2014.56 GOPS
  4   46.78 ms run,   13.41 ms python,   33.37 ms HSA * 6, 1161.00 loss, 0.000214 LR, 6.64 GB used,  43068.90 GFLOPS,   2014.56 GOPS
  5   45.42 ms run,   12.30 ms python,   33.11 ms HSA * 6, 1156.00 loss, 0.000256 LR, 6.64 GB used,  44355.12 GFLOPS,   2014.56 GOPS
  6   45.32 ms run,   12.29 ms python,   33.03 ms HSA * 6, 1138.00 loss, 0.000299 LR, 6.64 GB used,  44451.15 GFLOPS,   2014.56 GOPS
  ```

nimlgen commented on tinygrad/realize.py:
```diff
@@ -172,6 +172,40 @@ def _is_padding_okay(buf:LazyBuffer, realizes:Set[LazyBuffer]) -> bool:
+  free: Deque[int] = deque([idx for idx,deg in in_degrees.items() if deg == 0])
+
+  while free:
+    node = free.popleft()
```
I am thinking about this logic, while it's looks simple and good on some schedules, but is it generally a good heuristic for the thing we trying to solve? In my understanding we want to hide transfer times. This does really primitive reordering, how efficient is it with other schedules?

uuuvn replied:
14.88 tok/s on llama 7b on 6 gpus (w/o ring, just this optimization) vs 15.10 on #3626

uuuvn replied:
```
(.venv) uuuvn@tiny10:~/src/tinygrad$ STEPS=10 GPUS=6 BS=720 HSA=1 python3 examples/hlb_cifar10.py
shuffling training dataset in 1287.12 ms (epoch=0)
  0 14875.71 ms run, 14871.37 ms python,    4.34 ms HSA * 6, 1198.74 loss, 0.001495 LR, 1.10 GB used,     63.62 GFLOPS,    946.39 GOPS
  1 1131.90 ms run, 1129.36 ms python,    2.54 ms HSA * 6, 1198.97 loss, 0.002989 LR, 6.68 GB used,    834.47 GFLOPS,    944.53 GOPS
  2   46.13 ms run,   13.32 ms python,   32.80 ms HSA * 6, 2044.32 loss, 0.003133 LR, 6.69 GB used,  20476.73 GFLOPS,    944.53 GOPS
  3   38.37 ms run,   11.75 ms python,   26.61 ms HSA * 6, 2233.89 loss, 0.002698 LR, 6.69 GB used,  24617.35 GFLOPS,    944.53 GOPS
  4   38.32 ms run,   11.66 ms python,   26.66 ms HSA * 6, 1795.34 loss, 0.002262 LR, 6.69 GB used,  24645.85 GFLOPS,    944.53 GOPS
  5   38.27 ms run,   11.69 ms python,   26.59 ms HSA * 6, 1633.46 loss, 0.001827 LR, 6.69 GB used,  24679.91 GFLOPS,    944.53 GOPS
  6   38.61 ms run,   11.63 ms python,   26.97 ms HSA * 6, 1630.74 loss, 0.001392 LR, 6.69 GB used,  24466.32 GFLOPS,    944.53 GOPS
  7   38.44 ms run,   11.67 ms python,   26.77 ms HSA * 6, 1364.76 loss, 0.000956 LR, 6.69 GB used,  24572.66 GFLOPS,    944.53 GOPS
  8   38.51 ms run,   11.59 ms python,   26.92 ms HSA * 6, 1406.93 loss, 0.000521 LR, 6.69 GB used,  24525.51 GFLOPS,    944.53 GOPS
  9   38.40 ms run,   11.70 ms python,   26.70 ms HSA * 6, 1381.90 loss, 0.000086 LR, 6.69 GB used,  24594.88 GFLOPS,    944.53 GOPS
shuffling test dataset in 123.60 ms (epoch=0)
eval     1376/10080 13.65%,    2.49 val_loss STEP=10 (in 1206.81 ms)
```
hlb_cifar

nimlgen replied:
I mean results look good. But if we get a different initial schedule, is the heuristic still as efficient? If we have 4 available kernels to execute and last of them has a long copy child, this won't hide this long copy time, since it will execute other kernels, before execution 4th to start copy in parallel.

nimlgen commented:
profiled llama, seems that delay between packets in sdma queues is slightly higher than aql. I am going to measure the performance impact of moving some copies to aql queues

uuuvn commented:
![image](https://github.com/tinygrad/tinygrad/assets/83587632/f4548a36-6560-493b-a9dc-99b0364beac8)

uuuvn commented:
39 ms down from 47 ms on hlb_cifar with ring

uuuvn commented:
LLaMA slowness is fixed. Ready to review i think

But #3560 is required for benchmark on HSA not to crash and there is a over 6500 lines issue

chenyuxyz commented:
3560 is merged. Is this ready? I see cifar is slightly faster, llama is the same.

`python test/external/external_benchmark_multitensor_allreduce.py`

```
...
ring_allreduce iter 9/10: 0.085602 sec 102.22 GFLOP/s 11.68 GB/s
ring_allreduce iter 10/10: 0.085636 sec 102.18 GFLOP/s 11.68 GB/s
Traceback (most recent call last):
  File "/home/chenyu/tinygrad/test/external/external_benchmark_multitensor_allreduce.py", line 63, in <module>
    main()
  File "/home/chenyu/tinygrad/test/external/external_benchmark_multitensor_allreduce.py", line 58, in main
    (naive_gflops, naive_gbs, naive_secs) = test(naive_allreduce, devs, N)
  File "/home/chenyu/tinygrad/test/external/external_benchmark_multitensor_allreduce.py", line 32, in test
    realize(_jitted(impl, ReduceOps.SUM, Tensor(MultiLazyBuffer(lbs, 0), device=devs)).lazydata.lbs)
  File "/home/chenyu/tinygrad/tinygrad/features/jit.py", line 115, in __call__
    for ji in self.jit_cache: ji.prg(cast(List[Buffer], ji.rawbufs), var_vals, wait=DEBUG>=2, jit=True)
  File "/home/chenyu/tinygrad/tinygrad/runtime/graph/hsa.py", line 130, in __call__
    elif i == 1: self.transfers[self.ji_to_transfer[j]][2] = input_rawbuffers[input_idx]._buf # src
IndexError: list index out of range
```

chenyuxyz commented on tinygrad/features/multi.py:
```diff
@@ -1,16 +1,43 @@
-  return [functools.reduce(lambda x,y: x.e(bop, y), [x.copy_to_device(lb.device) for x in lbs]) for lb in lbs]
+
+  n_lbs, dim = len(lbs), prod(lbs[0].shape)
+  # Ring allreduce doesn't provide a benefit with only 2 nodes or where number of elements is less than 256k (empirically)
```
skipping 2 nodes makes sense. Can you show how you derived the 256k number.

Also what is expected to hit in CI benchmark? I see an improvement in `test/external/external_benchmark_multitensor_allreduce.py` and cifar training. Does it hit llama or resnet?

uuuvn replied:
> Can you show how you derived the 256k number.

By running it with different sizes and seeing what's faster

> Also what is expected to hit in CI benchmark? I see an improvement in test/external/external_benchmark_multitensor_allreduce.py and cifar training.

Should be +-12GB/s in benchmark

> Does it hit llama or resnet?

LLaMA no (small allreduce sizes), resnet idk
